### ISSUE TITLE
Init `PoolManager`

### ISSUE THREAD

**Issue by wischli** [centrifuge/protocol-v3#28]
* Create the `PoolManager` contract and add boilerplate for handling decoded incoming messages forwarded from the Gateway
   * Example: Gateway receives a message which decodes to `RequestDeposit { .. }` and calls `PoolManager.requestDeposit(..)`
   * NOTE: TBD if Gateway does encoding/decoding, definitely shouldn't be done by the `PoolManager` to reduce bloat
* Of course, not all parameters of the PoolManager handlers are fully clear -> decide on your own for which function declarations you want to leave out some/all parameters 
   E.g.
   ```solidity
  // If unsure
  function requestDeposit() { ... }
  
  // If somewhat clear
  function requestDeposit(uint64 poolId, bytes16 shareClassId, address investor, address accountingInvestorToken, uint128 amount)

* For now, error out with some `Unsupported` error for all incoming handler calls. We will incrementally add handlers for incoming messages in subsequent PRs
   E.g.
   ```solidity
   function requestDeposit(uint64 poolId, bytes16 shareClassId, address investor, address accountingInvestorToken, uint128 amount) {
      revert("Unsupported");
   }
   
* ~~Copy and paste messages from CV repository~~
* ~~Apply minor changes discussed during offsite: https://www.notion.so/LP-Integration-and-Changes-1342eac24e17802aa954c073a5b916e4~~

**Comment by lemunozm** [comment]
> Create the PoolManager contract and add a handler for all incoming CV messages

I would let the handler in the `Gateway`, and use normal methods in the `PoolManager`. My rationale:

- I think the `PoolManager` already does quite a job.
- It seems the `Gateway` duty to handle such messages and call to `PoolManager` in the form of i.e. `PoolManager.depositRequest()`.
- If new messages are added/removed/modified, they will not affect the `PoolManager` implementation because it's agnostic to the protocol. The Gateway takes care of them.

**Comment by wischli** [comment]
> > Create the PoolManager contract and add a handler for all incoming CV messages
> 
> I would let the handler in the `Gateway`, and use normal methods in the `PoolManager`. My rationale: [...]

Fair point, thanks for bringing that up! Do you think it makes sense to start the `PoolManager` impl without having a boilerplate Gateway or do you consider that to be a prerequisite?

**Comment by lemunozm** [comment]
> Do you think it makes sense to start the PoolManager impl without having a boilerplate Gateway

I think so. I see the `Gateway` as another actor, similar to the "FundManager", that, for now, we can emulate or mocked. Once we have a more solid `PoolManager` implementation, we can start with the `Gateway`.

A couple of things we should take care about the "Gateway" vs "FundManager" actors:
- This "Gateway" actor seen from `PoolManager` is a bit different from the "FundManager" actor. The "FundManager" can only unlock a pool if they is allowed for that pool (mostly the creator of the pool), but "Gateway" can unlock any pool.
- The "FundManager" should not be able to call `PoolManager.depositRequest()` to not break the invariants we do in CP against CV. So we need to protect those calls done by the "Gateway" actor to only be called by him

**Comment by lemunozm** [comment]
> The "FundManager" should not be able to call PoolManager.depositRequest() to not break the invariants we do in CP against CV. So we need to protect those calls done by the "Gateway" actor to only be called by him.

Maybe this is a contra-argument against my rationale above. Seems easy to check it if we have only one `handle(msg)` method in the `PoolManager` that the "Gateway" can call.

**Comment by wischli** [comment]
Idea: Simple contract with the `unlock` mechanism using transient storage. Apart from that no other storage because that would make migrations difficult.

**Comment by lemunozm** [comment]
I was just thinking about it!

**Comment by lemunozm** [comment]
> Idea: Simple contract with the unlock mechanism using transient storage. Apart from that no other storage because that would make migrations difficult.

I think, sadly, I can not do it easily. The unlock should check the `msg.sender` is an account able to act as a fund manager, so I would need the `PoolRegisty` dependency to check that. But requiring such dependency I think then I makes no sense to extract this unlocking from the `PoolManager`.

TLDL: The unlock logic should be in `PoolManager`.
